# Importing Libraries and Data

In [1]:
# Importing Libraries
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from datetime import datetime as dt

In [2]:
# Importing Data
Path = r'D:\Data_Analysis\05-12-2025_Bike_Dashboard\02.Data'
df_bikes = pd.read_pickle(os.path.join(Path, 'Prepared Data', 'Bike_Trips_Clean.pkl'))
df_bikes.head()

,ride_id,rideable_type,started_at,ended_at,start_station_name,start_station_id,end_station_name,end_station_id,start_lat,start_lng,end_lat,end_lng,member_casual,date,avgTemp,daily_rides,trip_duration
0,115C78C3039FFA89,electric_bike,2022-01-01 09:21:14,2022-01-01 09:35:46,Essex Light Rail,JC038,Essex Light Rail,JC038,40.712774,-74.036486,40.712774,-74.036486,member,2022-01-01,11.6,592,872.0
1,7FFD810CAA7A919E,classic_bike,2022-01-01 02:43:56,2022-01-01 02:43:57,12 St & Sinatra Dr N,HB201,12 St & Sinatra Dr N,HB201,40.750604,-74.024020,40.750604,-74.024020,member,2022-01-01,11.6,592,1.0
2,E715E8432031B72C,classic_bike,2022-01-01 02:13:33,2022-01-01 02:18:42,Essex Light Rail,JC038,Washington St,JC098,40.712774,-74.036486,40.724294,-74.035483,member,2022-01-01,11.6,592,309.0
3,BF1B7B1E1961A87B,electric_bike,2022-01-01 17:18:46,2022-01-01 18:55:25,Grand St,JC102,W 27 St & 7 Ave,6247.06,40.715178,-74.037683,40.746647,-73.993915,casual,2022-01-01,11.6,592,5799.0
4,4A01F0E53C6F4386,electric_bike,2022-01-01 11:23:32,2022-01-01 11:29:27,Christ Hospital,JC034,Hoboken Terminal - Hudson St & Hudson Pl,HB101,40.734786,-74.050444,40.735938,-74.030305,member,2022-01-01,11.6,592,355.0


# Preparing Data for Dashboard

In [3]:
# Creating a month column
df_bikes['month'] = df_bikes['date'].dt.month
df_bikes['month'] = df_bikes['month'].astype('int')
df_bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 895475 entries, 0 to 895474
Data columns (total 18 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             895475 non-null  object        
 1   rideable_type       895475 non-null  object        
 2   started_at          895475 non-null  datetime64[ns]
 3   ended_at            895475 non-null  datetime64[ns]
 4   start_station_name  895475 non-null  object        
 5   start_station_id    895475 non-null  object        
 6   end_station_name    892281 non-null  object        
 7   end_station_id      892281 non-null  object        
 8   start_lat           895475 non-null  float64       
 9   start_lng           895475 non-null  float64       
 10  end_lat             893505 non-null  float64       
 11  end_lng             893505 non-null  float64       
 12  member_casual       895475 non-null  object        
 13  date                895475 no

In [4]:
# Creating a season column
df_bikes['season'] = ['winter' if (month == 12 or 1 <= month <= 4)
                     else 'spring' if (4 < month <= 5)
                     else 'summer' if (5 < month <= 9)
                     else 'fall'
                     for month in df_bikes['month']]
df_bikes.tail()

,ride_id,rideable_type,started_at,ended_at,start_station_name,start_station_id,end_station_name,end_station_id,start_lat,start_lng,end_lat,end_lng,member_casual,date,avgTemp,daily_rides,trip_duration,month,season
895470,1C2DA98EA3E22C36,electric_bike,2022-12-31 11:04:03,2022-12-31 11:07:05,Southwest Park - Jackson St & Observer Hwy,HB401,City Hall - Washington St & 1 St,HB105,40.737551,-74.041664,40.737360,-74.030970,casual,2022-12-31,8.2,983,182.0,12,winter
895471,15CAD9793CBB48C2,classic_bike,2022-12-31 09:58:06,2022-12-31 10:11:04,Riverview Park,JC057,Manila & 1st,JC082,40.744703,-74.043759,40.721651,-74.042884,casual,2022-12-31,8.2,983,778.0,12,winter
895472,D48F644880B431D1,classic_bike,2022-12-31 16:35:25,2022-12-31 17:00:09,Manila & 1st,JC082,Riverview Park,JC057,40.721651,-74.042884,40.744319,-74.043991,casual,2022-12-31,8.2,983,1484.0,12,winter
895473,1E992E823B5058B7,classic_bike,2022-12-31 16:05:25,2022-12-31 16:12:24,City Hall - Washington St & 1 St,HB105,11 St & Washington St,HB502,40.737360,-74.030970,40.749985,-74.027150,casual,2022-12-31,8.2,983,419.0,12,winter
895474,055D6A3EC0CD3E03,classic_bike,2022-12-31 18:00:20,2022-12-31 18:11:33,11 St & Washington St,HB502,City Hall - Washington St & 1 St,HB105,40.749985,-74.027150,40.737360,-74.030970,member,2022-12-31,8.2,983,673.0,12,winter


In [5]:
# Creating days of the week columns
df_bikes['day_name'] = df_bikes['started_at'].dt.day_name()
df_bikes['time_period'] = pd.cut(df_bikes['started_at'].dt.hour, bins=[0,5,12,17,21,24], labels=['Night', 'Morning', 'Afternoon', 'Evening', 'Night'], right=False, ordered=False)
df_bikes.head(10)

,ride_id,rideable_type,started_at,ended_at,start_station_name,start_station_id,end_station_name,end_station_id,start_lat,start_lng,...,end_lng,member_casual,date,avgTemp,daily_rides,trip_duration,month,season,day_name,time_period
0,115C78C3039FFA89,electric_bike,2022-01-01 09:21:14,2022-01-01 09:35:46,Essex Light Rail,JC038,Essex Light Rail,JC038,40.712774,-74.036486,...,-74.036486,member,2022-01-01,11.6,592,872.0,1,winter,Saturday,Morning
1,7FFD810CAA7A919E,classic_bike,2022-01-01 02:43:56,2022-01-01 02:43:57,12 St & Sinatra Dr N,HB201,12 St & Sinatra Dr N,HB201,40.750604,-74.024020,...,-74.024020,member,2022-01-01,11.6,592,1.0,1,winter,Saturday,Night
2,E715E8432031B72C,classic_bike,2022-01-01 02:13:33,2022-01-01 02:18:42,Essex Light Rail,JC038,Washington St,JC098,40.712774,-74.036486,...,-74.035483,member,2022-01-01,11.6,592,309.0,1,winter,Saturday,Night
3,BF1B7B1E1961A87B,electric_bike,2022-01-01 17:18:46,2022-01-01 18:55:25,Grand St,JC102,W 27 St & 7 Ave,6247.06,40.715178,-74.037683,...,-73.993915,casual,2022-01-01,11.6,592,5799.0,1,winter,Saturday,Evening
4,4A01F0E53C6F4386,electric_bike,2022-01-01 11:23:32,2022-01-01 11:29:27,Christ Hospital,JC034,Hoboken Terminal - Hudson St & Hudson Pl,HB101,40.734786,-74.050444,...,-74.030305,member,2022-01-01,11.6,592,355.0,1,winter,Saturday,Morning
5,13C256ED54892996,classic_bike,2022-01-01 02:44:21,2022-01-01 02:49:51,12 St & Sinatra Dr N,HB201,Hoboken Terminal - River St & Hudson Pl,HB102,40.750604,-74.024020,...,-74.029127,member,2022-01-01,11.6,592,330.0,1,winter,Saturday,Night
6,164AC1EDDC26C06C,electric_bike,2022-01-01 01:25:10,2022-01-01 01:32:04,12 St & Sinatra Dr N,HB201,Hoboken Terminal - River St & Hudson Pl,HB102,40.750604,-74.024020,...,-74.029127,casual,2022-01-01,11.6,592,414.0,1,winter,Saturday,Night
7,991B41B69E86FBAA,classic_bike,2022-01-01 01:26:23,2022-01-01 01:31:32,Willow Ave & 12 St,HB505,Willow Ave & 12 St,HB505,40.751867,-74.030377,...,-74.030377,member,2022-01-01,11.6,592,309.0,1,winter,Saturday,Night
8,FB997C42FD9086C3,electric_bike,2022-01-01 13:53:05,2022-01-01 14:08:47,Bergen Ave & Stegman St,JC108,Bergen Ave & Stegman St,JC108,40.706575,-74.086701,...,-74.086701,casual,2022-01-01,11.6,592,942.0,1,winter,Saturday,Afternoon
9,4951439081EE64AB,electric_bike,2022-01-01 17:07:13,2022-01-01 17:22:58,Pershing Field,JC024,9 St HBLR - Jackson St & 8 St,HB305,40.742677,-74.051789,...,-74.038412,casual,2022-01-01,11.6,592,945.0,1,winter,Saturday,Evening


In [ ]:
histday = go.Figure(data=[go.Histogram(histfunc='count', x=df_bikes['day_name'], name='count')]).update_xaxes(categoryorder='total descending')
histday.show()

In [ ]:
dayProb = df_bikes.groupby(['day_name', 'time_period'], observed=False)['member_casual'].apply(lambda x: (x=='member').mean()).reset_index(name='P(MemberperDay)')
day_prob = go.Figure(data=go.Heatmap(z=dayProb['P(MemberperDay)'], x=dayProb['day_name'], y=dayProb['time_period'], hoverongaps=False, texttemplate='%{z}', textfont={'size':10}))
day_prob.show()

In [ ]:
# memberSProb = df_bikes.groupby('season')['member_casual'].apply(lambda x: (x=='member').mean()).reset_index(name='P(MemberperSeason)')
memberProb = df_bikes.groupby('day_name')['member_casual'].apply(lambda x: (x=='member').mean()).reset_index(name='P(MemberperDay)')
member_prob = go.Figure(go.Bar(x=memberProb['day_name'], y=memberProb['P(MemberperDay)'], marker={'color':memberProb['P(MemberperDay)'], 'colorscale':'Blues'}))
member_prob.update_layout(title=dict(text='Member Probability of Riding per Day', x=0.5, font=dict(size=18, color='navy', family='bree, sans-serif')),
                 xaxis_title = 'Day of the Week',
                 xaxis = dict(color='navy', categoryorder='total ascending', title=dict(font=dict(size=14))),
                 yaxis_title = 'Member Probability of Riding',
                 yaxis = dict(gridcolor='rgba(0,0,0,0.1)', zerolinecolor='rgba(0,0,0,0.1)', color='navy', title=dict(font=dict(size=14))),
                 plot_bgcolor='lightsteelblue',
                 paper_bgcolor='white',
                 width = 900, height=600)
member_prob.show()

In [ ]:
# Using Graph Objects to plot Trip Duration line chart
durline = df_bikes.groupby('day_name')['trip_duration'].mean().reset_index(name='trip_duration')
duration_line = go.Figure(data=go.Scatter(x=durline['day_name'], y=durline['trip_duration'], name='Trip Duration', marker={'color':durline['day_name'], 'color':'navy'}))

# Find the max and round up to nearest minute mark
end_point = (max(df_bikes['trip_duration'])//60 + 1) * 60
tick_loc = np.arange(0, end_point + 1, 180)
tick_labels = [f'{int(s // 60):02d}:{int(s % 60):02d}' for s in tick_loc]

# Updating layout and Axes
duration_line.update_layout(title=dict(text='CitiBike Trip Duration per Day of the Week', x=0.5, font=dict(size=18, color='navy', family='bree, sans-serif')),
                 yaxis_title = 'Duration (Minutes)',
                 yaxis = dict(tickmode='array', tickvals=tick_loc, ticktext=tick_labels, range=[0, 1500], title=dict(font=dict(size=14))),
                 plot_bgcolor='lightsteelblue',
                 paper_bgcolor='white',
                 width=1_800, height=800)
duration_line.show()

# Visualizations

In [ ]:
user_box = go.Figure()

for ride_type in df_bikes['rideable_type'].unique():
    user_box.add_trace(go.Box(x=df_bikes[df_bikes['rideable_type'] == ride_type]['member_casual'], y=df_bikes[df_bikes['rideable_type'] == ride_type]['avgTemp'], name=ride_type, boxmean=True))

user_box.update_layout(title=dict(text='Average Temperature per User Type', x=0.15, font=dict(size=18, color='navy', family='bree, sans-serif')),
                             legend=dict(yanchor='auto', y=0.4, xanchor='auto', x=0.26, bgcolor='rgba(255,255,255,0.5)'),
                             plot_bgcolor='lightsteelblue',
                             paper_bgcolor='white',
                             boxmode='group',
                             width=500, height=600)
user_box.update_xaxes(title_text='Type of Bike User', color='navy', title=dict(font=dict(size=16)))
user_box.update_yaxes(title_text='Daily Average Temperature', color='violet', title=dict(font=dict(size=16)))
user_box.show()

In [ ]:
user_box = go.Figure()

for ride_type in df_bikes['rideable_type'].unique():
    user_box.add_trace(go.Box(x=df_bikes[df_bikes['rideable_type'] == ride_type]['member_casual'], y=df_bikes[df_bikes['rideable_type'] == ride_type]['daily_rides'], name=ride_type, boxmean=True))

user_box.update_layout(title=dict(text='Trip Duration per User Type', x=0.15, font=dict(size=18, color='navy', family='bree, sans-serif')),
                             legend=dict(yanchor='auto', y=0.3, xanchor='auto', x=0.26, bgcolor='rgba(255,255,255,0.5)'),
                             plot_bgcolor='lightsteelblue',
                             paper_bgcolor='white',
                             boxmode='group',
                             width=500, height=600)
user_box.update_xaxes(title_text='Type of User', color='navy', title=dict(font=dict(size=16)))
user_box.update_yaxes(title_text='Trip Duration', color='violet', title=dict(font=dict(size=16)))
user_box.show()

## Top 20 Start Stations Bar Plot

In [ ]:
# Selecting the top 20 start stations per value counts
topStart = df_bikes['start_station_name'].value_counts().head(20)

# Creating Plotly bar chart
fig_top20 = go.Figure(go.Bar(y=topStart.index, x=topStart.values, orientation='h', marker={'color':topStart.values, 'colorscale':'Blues'}))
fig_top20.update_layout(title=dict(text='Most Popular CitiBike Stations in New York', x=0.5, font=dict(size=18, color='navy', family='bree, sans-serif')),
                 yaxis_title = 'Start Stations',
                 yaxis = dict(color='navy', categoryorder='total ascending', title=dict(font=dict(size=14))),
                 xaxis_title = 'Sum of Trips',
                 xaxis = dict(gridcolor='rgba(0,0,0,0.1)', zerolinecolor='rgba(0,0,0,0.1)', color='navy', title=dict(font=dict(size=14))),
                 plot_bgcolor='lightsteelblue',
                 paper_bgcolor='white',
                 width = 900, height=600)
fig_top20.show()

# Calculating trip percentage of total 20 station trips captured by the top 5 stations.
Top4_proportion = (df_bikes['start_station_name'].value_counts().head(4).sum()) / \
    (df_bikes['start_station_name'].value_counts().head(20).sum())
print(f'\033[1mThe Top 4 stations account for {Top4_proportion:.2%} of the top 20 station trips.\033[0m')

## Daily Rides and Average Temperature Dual Axis Line Plot

In [ ]:
# Using Graph Objects to plot dual axis line chart
fig_DAline = make_subplots(specs = [[{'secondary_y':True}]])
fig_DAline.add_trace(go.Scatter(x=df_bikes['date'], y=df_bikes['daily_rides'], name='Daily Rides', marker={'color':df_bikes['daily_rides'], 'color':'navy'}),
               secondary_y=False)
fig_DAline.add_trace(go.Scatter(x=df_bikes['date'], y=df_bikes['avgTemp'], name='Daily Average Temperatures', marker={'color':df_bikes['avgTemp'], 'color':'violet'}),
               secondary_y=True)

# Updating layout and Axes
fig_DAline.update_layout(title=dict(text='CitiBike Daily Rides and Average Temperature (2022)', x=0.5, font=dict(size=18, color='navy', family='bree, sans-serif')),
                 plot_bgcolor='lightsteelblue',
                 paper_bgcolor='white',
                 width=1800, height=800)
fig_DAline.update_yaxes(title_text='Sum of Daily Rides', secondary_y=False, color='navy', title=dict(font=dict(size=14)))
fig_DAline.update_yaxes(title_text='Daily Average Temperature', secondary_y=True, color='violet', title=dict(font=dict(size=14)))
fig_DAline.show()

# Calculating Pearson's correlation coefficient
correlation = df_bikes['daily_rides'].corr(df_bikes['avgTemp'])
print(f'Daily Rides and Daily Average Temperature have a {"Strong Positive" if correlation > 0.7 else "Moderate Positive" if correlation > 0.4 else "Weak"} Correlation: {correlation:.3f}')

# Reducing size of Top 20 Station and Bikes DataFrame

In [14]:
# Changing data type of 20 Stations DataFrame
topStart = pd.DataFrame(topStart).reset_index()
topStart.rename(columns={'start_station_name':'start_station', 'count':'total_trips'}, inplace=True)
topStart['total_trips'] = topStart['total_trips'].astype('uint16')
topStart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   start_station  20 non-null     object
 1   total_trips    20 non-null     uint16
dtypes: object(1), uint16(1)
memory usage: 328.0+ bytes


In [15]:
# Keeping only essential columns of df_bikes data frame
bikes_red = df_bikes[['start_station_name', 'end_station_name', 'member_casual', 'avgTemp', 'season', 'date', 'day_name', 'time_period', 'rideable_type', 'daily_rides', 'trip_duration']].copy()
bikes_red['daily_rides'] = bikes_red['daily_rides'].astype('int16')
bikes_red.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 895475 entries, 0 to 895474
Data columns (total 11 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   start_station_name  895475 non-null  object        
 1   end_station_name    892281 non-null  object        
 2   member_casual       895475 non-null  object        
 3   avgTemp             895475 non-null  float64       
 4   season              895475 non-null  object        
 5   date                895475 non-null  datetime64[ns]
 6   day_name            895475 non-null  object        
 7   time_period         895475 non-null  category      
 8   rideable_type       895475 non-null  object        
 9   daily_rides         895475 non-null  int16         
 10  trip_duration       895475 non-null  float64       
dtypes: category(1), datetime64[ns](1), float64(2), int16(1), object(6)
memory usage: 64.0+ MB


In [16]:
# Creating a Random Split of Bikes Reduced DataFrame
np.random.seed(42)
red = np.random.rand(len(bikes_red)) <= 0.92
small = bikes_red[~red]
small.tail()

,start_station_name,end_station_name,member_casual,avgTemp,season,date,day_name,time_period,rideable_type,daily_rides,trip_duration
895430,Adams St & 2 St,Bloomfield St & 15 St,member,8.2,winter,2022-12-31,Saturday,Evening,classic_bike,983,410.0
895439,City Hall - Washington St & 1 St,Clinton St & 7 St,member,8.2,winter,2022-12-31,Saturday,Afternoon,classic_bike,983,225.0
895443,City Hall - Washington St & 1 St,Clinton St & 7 St,casual,8.2,winter,2022-12-31,Saturday,Morning,classic_bike,983,1425.0
895453,4 St & Grand St,11 St & Washington St,casual,8.2,winter,2022-12-31,Saturday,Evening,classic_bike,983,219.0
895459,City Hall - Washington St & 1 St,11 St & Washington St,member,8.2,winter,2022-12-31,Saturday,Evening,classic_bike,983,670.0


In [17]:
# Reviewing statistics to confirm data integrity
small.describe()

,avgTemp,date,daily_rides,trip_duration
count,71811.000000,71811,71811.000000,71811.000000
mean,17.525364,2022-07-21 19:27:50.735680768,2947.094860,1046.337971
min,-11.700000,2022-01-01 00:00:00,111.000000,-3331.000000
25%,11.600000,2022-05-22 00:00:00,2214.000000,271.000000
50%,19.200000,2022-07-27 00:00:00,3188.000000,431.000000
75%,24.300000,2022-09-27 00:00:00,3686.000000,737.000000
max,31.300000,2022-12-31 00:00:00,4527.000000,787327.000000
std,8.476559,NaN,937.039412,6869.422085


# Actionable Insights

## How much to scale back between November and April?

In [18]:
# Creating Bike-Type Monthly Fleet Planning Table based on a total fleet of 8,000 of the 38,000 bikes available
fleet_size = 8_000
scaling_months = [11,12,1,2,3,4]

demand_type = df_bikes[df_bikes['rideable_type'].isin(['classic_bike', 'electric_bike'])]
demand_type = (demand_type.groupby([demand_type['date'].dt.month, 'rideable_type'])['daily_rides'].mean().reset_index().rename(columns={'date':'Month' ,'daily_rides':'Average Daily Rides'}))
peaks = (demand_type.groupby('rideable_type')['Average Daily Rides'].max().rename('peak_demand').reset_index())
demand_type = demand_type.merge(peaks, on='rideable_type')
demand_type['demand_rate'] = demand_type['Average Daily Rides']/demand_type['peak_demand']

In [19]:
fleet = peaks.assign(fleet=lambda x: x['peak_demand']/x['peak_demand'].sum())
fleet

,rideable_type,peak_demand,fleet
0,classic_bike,3748.655241,0.500077
1,electric_bike,3747.494567,0.499923


In [20]:
# Allocating fleet share by bike type
demand_type = demand_type.merge(fleet[['rideable_type', 'fleet']], on='rideable_type')
demand_type.rename(columns={'rideable_type':'Bike Type'}, inplace=True)
demand_type['Planned Fleet'] = (fleet_size * demand_type['fleet'] * demand_type['demand_rate']).round().astype(int)
demand_type

,Month,Bike Type,Average Daily Rides,peak_demand,demand_rate,fleet,Planned Fleet
0,1,classic_bike,969.520824,3748.655241,0.258632,0.500077,1035
1,1,electric_bike,955.897359,3747.494567,0.255076,0.499923,1020
2,2,classic_bike,1282.614859,3748.655241,0.342153,0.500077,1369
3,2,electric_bike,1250.469120,3747.494567,0.333681,0.499923,1335
4,3,classic_bike,1816.312134,3748.655241,0.484524,0.500077,1938
5,3,electric_bike,1794.790972,3747.494567,0.478931,0.499923,1915
6,4,classic_bike,2221.209199,3748.655241,0.592535,0.500077,2371
7,4,electric_bike,2221.966259,3747.494567,0.592920,0.499923,2371
8,5,classic_bike,2814.583447,3748.655241,0.750825,0.500077,3004
9,5,electric_bike,2762.493832,3747.494567,0.737158,0.499923,2948


In [21]:
# Defining period and reduction metrics
demand_type['Period'] = demand_type['Month'].apply(lambda sm: 'Scaling' if sm in scaling_months else 'Peak')
demand_type['Fleet Reduction'] = ((fleet_size * demand_type['fleet']).round().astype(int)) - (demand_type['Planned Fleet'])
demand_type

,Month,Bike Type,Average Daily Rides,peak_demand,demand_rate,fleet,Planned Fleet,Period,Fleet Reduction
0,1,classic_bike,969.520824,3748.655241,0.258632,0.500077,1035,Scaling,2966
1,1,electric_bike,955.897359,3747.494567,0.255076,0.499923,1020,Scaling,2979
2,2,classic_bike,1282.614859,3748.655241,0.342153,0.500077,1369,Scaling,2632
3,2,electric_bike,1250.469120,3747.494567,0.333681,0.499923,1335,Scaling,2664
4,3,classic_bike,1816.312134,3748.655241,0.484524,0.500077,1938,Scaling,2063
5,3,electric_bike,1794.790972,3747.494567,0.478931,0.499923,1915,Scaling,2084
6,4,classic_bike,2221.209199,3748.655241,0.592535,0.500077,2371,Scaling,1630
7,4,electric_bike,2221.966259,3747.494567,0.592920,0.499923,2371,Scaling,1628
8,5,classic_bike,2814.583447,3748.655241,0.750825,0.500077,3004,Peak,997
9,5,electric_bike,2762.493832,3747.494567,0.737158,0.499923,2948,Peak,1051


In [22]:
demand_type['Reduction Percent'] = (demand_type['Fleet Reduction']/(fleet_size * demand_type['fleet'])).round(4)*100
demand_type

,Month,Bike Type,Average Daily Rides,peak_demand,demand_rate,fleet,Planned Fleet,Period,Fleet Reduction,Reduction Percent
0,1,classic_bike,969.520824,3748.655241,0.258632,0.500077,1035,Scaling,2966,74.14
1,1,electric_bike,955.897359,3747.494567,0.255076,0.499923,1020,Scaling,2979,74.49
2,2,classic_bike,1282.614859,3748.655241,0.342153,0.500077,1369,Scaling,2632,65.79
3,2,electric_bike,1250.469120,3747.494567,0.333681,0.499923,1335,Scaling,2664,66.61
4,3,classic_bike,1816.312134,3748.655241,0.484524,0.500077,1938,Scaling,2063,51.57
5,3,electric_bike,1794.790972,3747.494567,0.478931,0.499923,1915,Scaling,2084,52.11
6,4,classic_bike,2221.209199,3748.655241,0.592535,0.500077,2371,Scaling,1630,40.74
7,4,electric_bike,2221.966259,3747.494567,0.592920,0.499923,2371,Scaling,1628,40.71
8,5,classic_bike,2814.583447,3748.655241,0.750825,0.500077,3004,Peak,997,24.92
9,5,electric_bike,2762.493832,3747.494567,0.737158,0.499923,2948,Peak,1051,26.28


In [23]:
# Sorting values of final table
fleet_plan = demand_type[demand_type['Period'] == 'Scaling']
fleet_plan = fleet_plan.sort_values(['Month', 'Bike Type'])[['Month', 'Bike Type', 'Average Daily Rides', 'Planned Fleet', 'Fleet Reduction', 'Reduction Percent']]
fleet_plan

,Month,Bike Type,Average Daily Rides,Planned Fleet,Fleet Reduction,Reduction Percent
0,1,classic_bike,969.520824,1035,2966,74.14
1,1,electric_bike,955.897359,1020,2979,74.49
2,2,classic_bike,1282.614859,1369,2632,65.79
3,2,electric_bike,1250.469120,1335,2664,66.61
4,3,classic_bike,1816.312134,1938,2063,51.57
5,3,electric_bike,1794.790972,1915,2084,52.11
6,4,classic_bike,2221.209199,2371,1630,40.74
7,4,electric_bike,2221.966259,2371,1628,40.71
20,11,classic_bike,2636.161667,2813,1188,29.70
21,11,electric_bike,2719.126090,2902,1097,27.43


Since Citibike currently counts with approximately 20,000 classic bikes and 15,000 electric bike (according to The City LifeOrg) in New York City and New Jersey, the fleet size was calculated based on a total of 8,000 bikes divided by bike type according to the usage revealed in the dataset. Since monthly demand fluctuates substantially, a month by month approach was selected to define the scaling strategy. The above table quantifies monthly fleet numbers that can be removed safely based on daily ride demand per bike type with a margin to maintain service levels. Considering existing demand, both bike types experience the same level of seasonal expansion in December and contraction in April that decreases further in the following months. From it, we can infer that November service levels can be safely maintained with an approximate 70% fleet size based on August peak numbers. It's suggested to schedule maintenance and bike substitution according to the monthly fleet reduction numbers.

## How many more stations to add along the water?

In [24]:
# Combining all stations into one column to select unique values only
all_stations = pd.melt(df_bikes, id_vars=['ride_id'], value_vars=['start_station_name', 'end_station_name'], var_name='station_type', value_name='station_name')
total_stations = all_stations['station_name'].nunique()
current_Wstations = 21

# Trip Percent hoarded by waterfront stations
trip_share = (df_bikes['ride_id'].loc[(df_bikes['start_station_name'].isin(['Morris Canal', 'Essex Light Rail', 'Paulus Hook', 'Columbus Dr at Exchange Pl', 'Harbor Side', 'Grant St', 'Washington St.', 'Newport PATH', 'Newport Pkwy', 'Hoboken Terminal – River St & Hudson Pl', 'Hoboken Terminal – Hudson St & Hudson Pl', 'South Waterfront Walkway – Sinatra Dr & 1 St', 'Hudson St & 4 St', 'Stevens – River Ter & 6 St', '12 St & Sinatra Dr N', '11 St & Washington St', '14 St Ferry - 14 St & Shipyard Ln', 'Bloomfield St & 15 St', 'Liberty Light Rail', 'JC Medical Center', 'Marin Light Rail'])) | (df_bikes['end_station_name'].isin(['Morris Canal', 'Essex Light Rail', 'Paulus Hook', 'Columbus Dr at Exchange Pl', 'Harbor Side', 'Grant St', 'Washington St.', 'Newport PATH', 'Newport Pkwy', 'Hoboken Terminal – River St & Hudson Pl', 'Hoboken Terminal – Hudson St & Hudson Pl', 'South Waterfront Walkway – Sinatra Dr & 1 St', 'Hudson St & 4 St', 'Stevens – River Ter & 6 St', '12 St & Sinatra Dr N', '11 St & Washington St', '14 St Ferry - 14 St & Shipyard Ln', 'Bloomfield St & 15 St', 'Liberty Light Rail', 'JC Medical Center', 'Marin Light Rail']))]).count()/df_bikes['ride_id'].count()
print(f'{trip_share:.2%}')

37.12%


In [25]:
# Calculating stations to add metrics
station_share = current_Wstations/total_stations
busy_rate = trip_share/station_share

# Calculating number of water sations to add along water per multiplier
multiplier = [1, 2, 3]
rows = []
for m in multiplier:
    requiredStations = ((trip_share * total_stations) / m).round()
    stationToAdd = max(0, requiredStations - current_Wstations)
    rows.append({
        'Load Multiplier': m,
        'Required Water Stations': requiredStations,
        'Additional Water Stations': stationToAdd})

# Creating a Descriptive Summary
summary = {
    'Total Stations': total_stations,
    'Existing Water Stations': current_Wstations,
    'Existing Water Stations Share of Total Stations': f'{station_share:.2%}',
    'Trip Share of Existing Water Stations': f'{trip_share:.2%}'}

print('\033[1mSummary\033[0m')
for k, v in summary.items():
    print(f'{k:50s}: {v}')

df_stations_to_add = pd.DataFrame(rows)
print('\nTarget Table with Stations to add along the water.')
print(df_stations_to_add.to_string(index=False))

Summary
Total Stations                                    : 320
Existing Water Stations                           : 21
Existing Water Stations Share of Total Stations   : 6.56%
Trip Share of Existing Water Stations             : 37.12%

Target Table with Stations to add along the water.
 Load Multiplier  Required Water Stations  Additional Water Stations
               1                    119.0                       98.0
               2                     59.0                       38.0
               3                     40.0                       19.0


A total of 21 water front stations were identified in the dataset based on the premise that waterfront station are those located within 300 meters from the water. With it, it was evident that the total share of stations of approximately 6.56% accounts for over 1/3 of the total trips meaning that each station received approximately 5.6 more trips than the average station. The above results reflect the number of stations that can be relocated along the water to keep the total stations the same (320) with a load multiplier that safely reduces the existing 5x load to a maximum of 3x. The required water stations at a load of 3x would be 40 from which the 21 identified water stations can be subtracted, which would leave us with a total stations to relocate as additional water stations of 19. Start with a targeted pilot plan of 19 stations to relocate from spatially close non-water stations with low activity and increase it to 98 stations progressively so as to not affect existing non-water stations.

In [26]:
# Converting results to one dataframe
# df_summary = (pd.DataFrame(list(summary.items()), columns=['metric', 'value']).assign(section='summary'))
# df_stations_to_add['section'] = 'scenario_options'
# df_stations_to_add = df_stations_to_add[['section'] + [c for c in df_stations_to_add.columns if c != 'section']]
# combined = pd.concat([df_summary, df_stations_to_add], ignore_index=True, sort=False)
# combined

In [27]:
# Converting results to one dataframe
df_summary = (pd.DataFrame(list(summary.items()), columns=['metric', 'value']).assign(section='summary'))
df_stations_to_add.insert(0, 'section', 'scenario_options')
combined = pd.concat([df_summary, df_stations_to_add], ignore_index=True, copy=False, sort=False)
combined

,metric,value,section,Load Multiplier,Required Water Stations,Additional Water Stations
0,Total Stations,320,summary,NaN,NaN,NaN
1,Existing Water Stations,21,summary,NaN,NaN,NaN
2,Existing Water Stations Share of Total Stations,6.56%,summary,NaN,NaN,NaN
3,Trip Share of Existing Water Stations,37.12%,summary,NaN,NaN,NaN
4,NaN,NaN,scenario_options,1.0,119.0,98.0
5,NaN,NaN,scenario_options,2.0,59.0,38.0
6,NaN,NaN,scenario_options,3.0,40.0,19.0


## How to ensure bikes are always stocked at the most popular stations?

In [28]:
# Determining which are most popular stations to ensure don't stock out.
station_demand = (df_bikes.groupby('start_station_name').agg(avg_daily_starts=('ride_id', 'count'), peark_hour_share=('time_period', lambda x: (x == 'Morning').mean())).reset_index().rename(columns={'start_station_name':'Start Station', 'avg_daily_starts':'Average Daily Starts', 'peark_hour_share':'Peak Hour Share'}))
ten_percent = station_demand['Average Daily Starts'].quantile(0.85)
popular_stations = station_demand[station_demand['Average Daily Starts'] >= ten_percent].copy()
popular_stations

,Start Station,Average Daily Starts,Peak Hour Share
1,12 St & Sinatra Dr N,17137,0.229737
2,14 St Ferry - 14 St & Shipyard Ln,17066,0.222255
22,City Hall - Washington St & 1 St,23289,0.133711
25,Columbus Dr at Exchange Pl,16731,0.119240
37,Grove St PATH,42556,0.145855
38,Hamilton Park,20477,0.365483
42,Hoboken Ave at Monmouth St,19501,0.303728
43,Hoboken Terminal - Hudson St & Hudson Pl,30244,0.170745
44,Hoboken Terminal - River St & Hudson Pl,33020,0.134373
61,Marin Light Rail,19230,0.358034


In [29]:
# Creating a Stocking Risk Score and Minimum Bike Number Column
Rides_per_Bike = 3.5
Buffer = 1.5
popular_stations['Stockout Risk'] = (popular_stations['Average Daily Starts'] * (1 + popular_stations['Peak Hour Share']))
popular_stations['Min Bikes Required'] = (popular_stations['Average Daily Starts']/(Rides_per_Bike*Buffer)).round().astype(int)
popular_stations

,Start Station,Average Daily Starts,Peak Hour Share,Stockout Risk,Min Bikes Required
1,12 St & Sinatra Dr N,17137,0.229737,21074.0,3264
2,14 St Ferry - 14 St & Shipyard Ln,17066,0.222255,20859.0,3251
22,City Hall - Washington St & 1 St,23289,0.133711,26403.0,4436
25,Columbus Dr at Exchange Pl,16731,0.119240,18726.0,3187
37,Grove St PATH,42556,0.145855,48763.0,8106
38,Hamilton Park,20477,0.365483,27961.0,3900
42,Hoboken Ave at Monmouth St,19501,0.303728,25424.0,3714
43,Hoboken Terminal - Hudson St & Hudson Pl,30244,0.170745,35408.0,5761
44,Hoboken Terminal - River St & Hudson Pl,33020,0.134373,37457.0,6290
61,Marin Light Rail,19230,0.358034,26115.0,3663


In [30]:
# Creating a Re-stocking rule
popular_stations = popular_stations.sort_values('Stockout Risk', ascending=False)
quantile75 = popular_stations['Stockout Risk'].quantile(0.75)

popular_stations['Re-Stock Rule'] = np.where(popular_stations['Stockout Risk'] > quantile75, 'Overnight and Early AM Re-Stocking', 'Overnight Re-Stocking')
popular_stations

,Start Station,Average Daily Starts,Peak Hour Share,Stockout Risk,Min Bikes Required,Re-Stock Rule
37,Grove St PATH,42556,0.145855,48763.0,8106,Overnight and Early AM Re-Stocking
75,South Waterfront Walkway - Sinatra Dr & 1 St,34245,0.152927,39482.0,6523,Overnight and Early AM Re-Stocking
44,Hoboken Terminal - River St & Hudson Pl,33020,0.134373,37457.0,6290,Overnight and Early AM Re-Stocking
43,Hoboken Terminal - Hudson St & Hudson Pl,30244,0.170745,35408.0,5761,Overnight Re-Stocking
38,Hamilton Park,20477,0.365483,27961.0,3900,Overnight Re-Stocking
69,Newport Pkwy,21961,0.233733,27094.0,4183,Overnight Re-Stocking
22,City Hall - Washington St & 1 St,23289,0.133711,26403.0,4436,Overnight Re-Stocking
61,Marin Light Rail,19230,0.358034,26115.0,3663,Overnight Re-Stocking
42,Hoboken Ave at Monmouth St,19501,0.303728,25424.0,3714,Overnight Re-Stocking
68,Newport PATH,19965,0.174656,23452.0,3803,Overnight Re-Stocking


The above analysis defines the most popular stations based on daily ride count and considers trip volume and peak hours to calculate the minimum number of bikes required with a conservative risk buffer of 1.5 that focuses on re-stocking reliability. The resulting table lists the stations that will require close monitoring to ensure they remain fully stocked before the start and throughout the business day. As can be seen, Grove St. PATH averages to over 42,550 daily starts, South Waterfront Walkway averages to over 34,200 and Hoboken Terminals dominate system usage with over 33,000 daily starts. The first 3 stations have a very high demand that will require overnight and early AM re-stocking to avoid depletion. On the other end, Hamilton Park and Marin Light Rail shows a demand time compression of 36.5% and 35.8% respectively, which suggests these can empty faster than the operations team filling potential. These stations can create system wide failures and must be monitored closely with advanced stocking measures to prevent shortages that would not be satisfied with reactive fixes.

In [31]:
combined

,metric,value,section,Load Multiplier,Required Water Stations,Additional Water Stations
0,Total Stations,320,summary,NaN,NaN,NaN
1,Existing Water Stations,21,summary,NaN,NaN,NaN
2,Existing Water Stations Share of Total Stations,6.56%,summary,NaN,NaN,NaN
3,Trip Share of Existing Water Stations,37.12%,summary,NaN,NaN,NaN
4,NaN,NaN,scenario_options,1.0,119.0,98.0
5,NaN,NaN,scenario_options,2.0,59.0,38.0
6,NaN,NaN,scenario_options,3.0,40.0,19.0


# Exporting Reduced Data

In [32]:
topStart.to_csv(os.path.join(Path, 'Prepared Data', 'Top_Start.csv'), index=False)
small.to_csv(os.path.join(Path, 'Prepared Data', 'Reduced_Trips.csv'), index=False)
fleet_plan.to_csv(os.path.join(Path, 'Prepared Data', 'Fleet_Plan.csv'), index=False)
popular_stations.to_csv(os.path.join(Path, 'Prepared Data', 'Restocking_Plan.csv'), index=False)
combined.to_csv(os.path.join(Path, 'Prepared Data', 'Water_Stations.csv'), index=False)